File: tree/lc1676.go
package tree

func lowestCommonAncestorIV(root *TreeNode, nodes []*TreeNode) *TreeNode {

	var foundNode *TreeNode

	var traverse func(*TreeNode) int

	traverse = func(root *TreeNode) int {

		foundCount := 0
		if root == nil {
			return 0
		}

		left := traverse(root.Left)
		right := traverse(root.Right)

		foundCount += left + right
		// check current node
		for _, node := range nodes {
			if node == root {
				foundCount += 1
				break
			}
		}

		if len(nodes) == foundCount && foundNode == nil {
			foundNode = root
		}

		return foundCount
	}

	traverse(root)

	return foundNode
}

File: tree/lc1676.py
class Solution(object):

    found = None

    def lowestCommonAncestor(self, root, nodes):
        nodes = set(nodes)
        self.traverse(root, nodes)
        return self.found

    def traverse(self, root, nodes):
        if not root:
            return 0
        
        count = 0
        left = self.traverse(root.left, nodes)
        right = self.traverse(root.right, nodes)

        count += left + right

        if root in nodes:
            count += 1

        if count == len(nodes):
            self.found = self.found or root
        
        return count
       
    def lowestCommonAncestorV2(self, root, nodes):
        nodes = set(nodes)
        return self.traverseV2(root, nodes)

    def traverseV2(self, root, nodes):
        if not root:
            return None

        if root in nodes:
            return root

        left = self.traverseV2(root.left, nodes)
        right = self.traverseV2(root.right, nodes)

        if left and right:
            return root

        return left or right
